ray-curator/api-design.md

```diff
@@ -0,0 +1,198 @@
+
+## Background
+
+### Current State
+- **Existing NeMo-Curator OSS:** Built on Dask for Text, Image, Synthetic Data Generation and Hard negative mining
+- **Cosmos Curate:** Built on Ray core with Cosmos Xenna for streaming map-style pipeline
+- **API Pattern:** Stages/modules accept distributed Dask dataframes as input
+
+### Why Ray?
+- **Unified backend** for all modalities (text, image, video)
+- **Better heterogeneous computing** support allowing interleaving of CPU and GPU stages
+- **Fractional GPU support** (multiple models per GPU) and multi-GPU support for larger models
+- **Enhanced autoscaling** for dynamic workloads
+
+### Why not Ray?
+- Limited support for distributed dataframes/arrays, especially distributed operations like groupby
+- Eager computation model (addressed through lazy evaluation at curator level)
+- Requires Curator to build a physical plan, and implement optimizations like task fusion
+## Design Principles
+
+### Task-Centric Architecture
+Unlike the previous dataset-level operations, the new design operates on individual **Tasks** - batches of data that flow through the pipeline. This enables:
+- Finer-grained control and monitoring
+- Better resource utilization
+
+### Map-style (Data-Parallel) Execution
+All stages are designed to be map-style on tasks, meaning they take task as input and produce task as output. This allows for easy parallelization and scaling.
+- We do not enforce 1-1 mapping between input and output tasks, but rather allow for multiple output tasks from a single input task and multiple input tasks from a single output task. More specifically, a stage applies a transformation from `X` to `Y`, where both `X` and `Y` can be `Task | list[Task] | None`.
+
+
+## Core Components
+
+1. **Tasks** : The fundamental unit of data that flows through the curation pipeline.
+2. **Stages** : The fundamental processing unit that takes Tasks as input and produces Tasks as output.
+3. **Pipelines** : A collection of stages that defines the complete processing workflow.
+4. **Executors** : The component that orchestrates the execution of the pipeline.
+
+### Tasks
+
+A **Task** is the fundamental unit of data that flows through the curation pipeline, representing a batch of input data for processing.
+
+#### Base Task Implementation
+
+```python
+@dataclass
+class Task(ABC, Generic[T]):
+    """Abstract base class for tasks in the pipeline."""
+    task_id: str
+    dataset_name: str
+    data: T
+    _stage_perf: list[StagePerfStats] = field(default_factory=list)
+    _metadata: dict[str, Any] = field(default_factory=dict)
+
+    @property
+    @abstractmethod
+    def num_items(self) -> int:
+        """Get the number of items in this task."""
+
+    @abstractmethod
+    def validate(self) -> bool:
+        """Validate the task data."""
+```
+
+#### Example Task Types
+
+#### DocumentTask
+For text-based curation pipelines:
+
+```python
+@dataclass
+class DocumentTask(Task[pd.DataFrame]):
```

Comment:
Shouldn't we refer to the actual implementation we have? `class DocumentBatch(Task[pa.Table | pd.DataFrame])`?